include/seastar/util/chunked_deque.hh

```diff
@@ -0,0 +1,733 @@
+    }
+    chunk* n = new_chunk();
+    n->prev = nullptr;
+    if (_front_chunk) {
+        _front_chunk->prev = n;
+        n->begin = n->end = items_per_chunk;
+    } else {
+        _back_chunk = n;
+        n->begin = n->end = items_per_chunk / 2;
+    }
+    n->next = _front_chunk;
+    _front_chunk = n;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::ensure_room_back() {
+    // If we don't have a back chunk or it's full, we need to create a new one
+    if (_back_chunk != nullptr && _back_chunk->end < items_per_chunk) {
+        return;
+    }
+    chunk* n = new_chunk();
+    n->next = nullptr;
+    if (_back_chunk) {
+        _back_chunk->next = n;
+        n->begin = n->end = 0;
+    } else {
+        _front_chunk = n;
+        n->begin = n->end = items_per_chunk / 2;
+    }
+    n->prev = _back_chunk;
+    _back_chunk = n;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+void
+chunked_deque<T, items_per_chunk, save_free_chunks>::undo_room_front() noexcept {
+    if (!_front_chunk->empty()) {
+        return;
+    }
+    chunk* old = _front_chunk;
+    _front_chunk = old->next;
+    if (_front_chunk) {
+        _front_chunk->prev = nullptr;
+    } else {
+        _back_chunk = nullptr;
+    }
+    --_nchunks;
+    delete old;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+void
+chunked_deque<T, items_per_chunk, save_free_chunks>::undo_room_back() noexcept {
+    if (!_back_chunk->empty()) {
+        return;
+    }
+    chunk* old = _back_chunk;
+    _back_chunk = old->prev;
+    if (_back_chunk) {
+        _back_chunk->next = nullptr;
+    } else {
+        _front_chunk = nullptr;
+    }
+    --_nchunks;
+    delete old;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+template <typename... Args>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::emplace_front(Args&&... args) {
+    ensure_room_front();
+    auto next = _front_chunk->begin - 1;
+    auto p = &_front_chunk->items[next].data;
+    try {
+        new (p) T(std::forward<Args>(args)...);
+    } catch(...) {
+        undo_room_front();
+        throw;
+    }
+    _front_chunk->begin = next;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::push_front(const T& data) {
+    ensure_room_front();
+    auto next = _front_chunk->begin - 1;
+    auto p = &_front_chunk->items[next].data;
+    try {
+        new (p) T(data);
+    } catch(...) {
+        undo_room_front();
+        throw;
+    }
+    _front_chunk->begin = next;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::push_front(T&& data) {
+    ensure_room_front();
+    auto next = _front_chunk->begin - 1;
+    auto p = &_front_chunk->items[next].data;
```

Comment:
I see you don't use mask(next) so I finally understand you do NOT use a circular buffer for the chunk. Maybe this makes sense if the case of just a single chunk (< 128 items) is not the common case. But I would have liked to see some explanation on what is the design because it's hard to guess.